### ISSUE TITLE
Support ProxyCommand using stdio redirection

### ISSUE THREAD

**Issue by skrysm** [PowerShell/Win32-OpenSSH#355]
The current version of the port doesn't support `ProxyCommand`. A host defined like this results in the following error message:

```
proxy connect is not supported in Windows
```

Judging from the code structure in [sshconnect.c](https://github.com/PowerShell/Win32-OpenSSH/blob/946ed7d29761cfe239f1b80f08920b88f2236f42/sshconnect.c#L194), it seems that this hasn't been implemented _yet_.

However, during my (limited) search, I haven't found any mention of this. So I wanted to create an explicit issue for this - to track the progress on this missing feature and so that other people can find this more easily.

**Comment by manojampalam** [comment]
Thanks for the follow up. ProxyCommand is not in scope for the initial integration. Updated here - https://github.com/PowerShell/Win32-OpenSSH/wiki/Project-Scope

**Comment by niemyjski** [comment]
Is this coming in 1.1?

**Comment by manojampalam** [comment]
This is not in the horizon yet. 

How do you intend to use it? To support proxycommand, OpenSSH uses passing file descriptors over domain sockets. This concept does not exist on Windows. Even if we provide an alternative to doing this, the "proxycommand" application need to be aware of that specific protocol.

**Comment by niemyjski** [comment]
We use it to proxy ssh commands over a jump box

```
Host test-data0
    User test
    IdentityFile ~/.ssh/test
    Hostname 10.0.0.6
    ProxyCommand ssh -q -W %h:%p test-kibana
```

**Comment by kcking** [comment]
We use `ProxyCommand` in [Krypton](krypt.co) to read signatures returned from the server and verify them on the phone. This feature would bring us much closer to full Windows support.

I believe OpenSSH just uses stdin/stdout to make this happen (though I'm not sure how this maps to Windows) and there aren't any domain sockets are involved. For @niemyjski's example, `ssh test-data0` first spins up `ssh -q -W %h:%p test-kibana` as a subprocess, where `%h:%p` becomes `10.0.0.6:22`. Then the parent `ssh test-data0` process routes outgoing traffic to `stdin` of the child process instead of sending it over the network. The end result is that an ssh handshake happens between `ssh test-kibana` and the `test-kibana` server, then a second ssh handshake happens between `ssh test-data0` and `10.0.0.6` through the first ssh session.

We have some [documentation](https://krypt.co/docs/ssh/using-a-bastion-host.html) on `ProxyCommand` that might be worth a read (it's not Krypton-specific).

**Comment by manojampalam** [comment]
It would have been simple if OpenSSH just used stdin/stdout to stream protocol data. But that's not the case, it gets sock descriptor from "proxycommand" using [ancillary data](http://www.oocities.org/margineantm/syspp/background/ancillary.html). This "sock" is subsequently used for protocol communication.

Let me know if I am missing something.

**Comment by kcking** [comment]
Check out the `ssh_proxy_connect` function [here](https://github.com/PowerShell/Win32-OpenSSH/blob/946ed7d29761cfe239f1b80f08920b88f2236f42/sshconnect.c#L222). This creates the proxycommand subprocess and redirects the traffic of the parent process' connection through the subprocess' stdin/stdout (using pin and pout).

I have implemented a proxycommand before in Go ([krssh](https://github.com/kryptco/kr/blob/master/krssh/krssh.go#L237)) and am happy to help with advice/code over here if you like.

**Comment by lfourquaux** [comment]
@manojampalam: You are missing something: ProxyCommand can be used in 2 different ways.

The default, and older, use of ProxyCommand uses a subprocess that communicates using stdin/stdout.  It would not be too difficult to implement using common Win32 APIs.

When ProxyUseFdpass is set to true, in addition to ProxyCommand, the helper command is expected to use fd passing with ancillary data.  This can be more efficient, but is less portable (even to different Unices).  This would certainly be harder to implement, and I would suggest to wait and see how the Windows AF_UNIX implementation evolves.  This comment:
https://blogs.msdn.microsoft.com/commandline/2017/12/19/af_unix-comes-to-windows/#comment-40465
suggests that fd passing is "on the radar".

Here are some relevant quotes from "man ssh_config":
>      ProxyCommand
>              Specifies the command to use to connect to the server.  The com‐
>              mand string extends to the end of the line, and is executed using
>              the user's shell ‘exec’ directive to avoid a lingering shell
>              process.
> 
>              Arguments to ProxyCommand accept the tokens described in the
>              TOKENS section.  The command can be basically anything, and
>              should read from its standard input and write to its standard
>              output.  It should eventually connect an sshd(8) server running
>              on some machine, or execute sshd -i somewhere.  Host key manage‐
>              ment will be done using the HostName of the host being connected
>              (defaulting to the name typed by the user).  Setting the command
>              to none disables this option entirely.  Note that CheckHostIP is
>              not available for connects with a proxy command.
> 
>              This directive is useful in conjunction with nc(1) and its proxy
>              support.  For example, the following directive would connect via
>              an HTTP proxy at 192.0.2.0:
> 
>                 ProxyCommand /usr/bin/nc -X connect -x 192.0.2.0:8080 %h %p
[snip]
>      ProxyUseFdpass
>              Specifies that ProxyCommand will pass a connected file descriptor
>              back to ssh(1) instead of continuing to execute and pass data.
>              The default is no.

**Comment by manojampalam** [comment]
hmm yes. I was missing the other option. Yes, it should be straight forward to enable it. I'll see if it could be done by next release.

**Comment by manojampalam** [comment]
[ssh.zip](https://github.com/PowerShell/Win32-OpenSSH/files/1669583/ssh.zip)
Try this out. Implementation is slightly different from that in Unix - proxycommand is launched directly instead of in user's shell (since I've seen cmd.exe interfering with redirected io)

**Comment by kcking** [comment]
@manojampalam could you post the source code on a branch? Would love to check it out!

**Comment by manojampalam** [comment]
Sure. Tentative changes are in https://github.com/PowerShell/openssh-portable/pull/268

**Comment by kcking** [comment]
Thanks for the quick turnaround! I just tried the changes -- I get one error for `ProxyCommand` and a different one for `ProxyJump`

I tested `ProxyCommand` with `ProxyCommand nc64.exe %h %p` set in my ssh config and using the netcat at https://eternallybored.org/misc/netcat/
I get the following error when running `ssh me.krypt.co`:
```PS C:\Users\Kevin\Downloads> .\ssh.exe -vvv me.krypt.co
OpenSSH_7.6p1, LibreSSL 2.6.2
debug1: Reading configuration data C:\\Users\\Kevin/.ssh/config
debug1: C:\\Users\\Kevin/.ssh/config line 1: Applying options for *
debug3: Failed to open file:C:\\ProgramData\\ssh/ssh_config error:2
debug1: Executing proxy command: exec C:\\Users\\Kevin\\Downloads\\netcat-win32-1.12\\nc64.exe me.krypt.co 22
debug3: spawning C:\\Users\\Kevin\\Downloads\\netcat-win32-1.12\\nc64.exe me.krypt.co 22
debug3: w32_getpeername ERROR: not sock :2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_rsa error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_rsa.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_rsa type -1
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_rsa-cert error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_rsa-cert.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_rsa-cert type -1
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_dsa error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_dsa.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_dsa type -1
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_dsa-cert error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_dsa-cert.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_dsa-cert type -1
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_ecdsa error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_ecdsa.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_ecdsa type -1
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_ecdsa-cert error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_ecdsa-cert.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_ecdsa-cert type -1
debug1: identity file C:\\Users\\Kevin/.ssh/id_ed25519 type 3
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_ed25519-cert error:2
debug3: Failed to open file:C:\\Users\\Kevin/.ssh/id_ed25519-cert.pub error:2
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\Kevin/.ssh/id_ed25519-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_7.6
```

With `ssh me.krypt.co` and `ProxyJump test.com` in my ssh config, I get the following error: `Bad stdio forwarding specification ''[me.krypt.co]:22''` which seems to come from https://github.com/PowerShell/Win32-OpenSSH/blame/d6ce465b1528802b38ec7ececabc7755cc1b89eb/ssh.c#L800

**Comment by manojampalam** [comment]
Dont see any issues in the above traces yet. Not sure what's wrong. I'll spend some time next week.

If you want to explore, you could write a simple app that forwards traffic from/to stdio <--> port and use it as a proxy to sshd.

**Comment by kcking** [comment]
The first issue I see is the line:
```
debug3: w32_getpeername ERROR: not sock :2
```
but I haven't had a chance to dig into how significant it might be or where that comes from. Currently the ssh session just hangs at that last line.

`nc`, or netcat actually is exactly the app you mentioned. It links `stdin`/`stdout` with the host/port in the arguments. This means that the ssh config line `ProxyCommand nc64.exe %h %p` is basically a no-op proxy command that passes all traffic to the same destination but throught netcat.

Let me know if there's anything else I can do to help debug!

**Comment by manojampalam** [comment]
That line about getpeername is expected since, ssh is talking over a pipe. 

I'll try with netcat and see if I can find anything obvious.

**Comment by manojampalam** [comment]
Tried with netcat and see the same behavior that you detailed. The traces indicate that ssh.exe is not receiving any output from proxycommand.
sshd traces indicate that it did receive the initial exchange from client - this proves that the input pipe from ssh to sshd via netcat is working

I then tried with the following entry:
ProxyCommand C:\Windows\System32\cmd.exe /c echo hello_there & sleep 5

and do see this in the traces now:
debug1: ssh_exchange_identification: hello_there

This proves that return pipe from "proxy" to ssh is working too. So, at this point, it looks like an issue with netcat unable to write back to stdout (provided by ssh.exe).
